core/core-mcp/src/main/java/ai/wanaku/core/mcp/util/LabelExpressionParser.java

```diff
@@ -0,0 +1,308 @@
+package ai.wanaku.core.mcp.util;
+
+import ai.wanaku.api.types.LabelsAwareEntity;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+/**
+ * Recursive descent parser for label expression filters.
+ * <p>
+ * Parses expressions like:
+ * <ul>
+ *   <li>"category=weather"</li>
+ *   <li>"category=weather & action=forecast"</li>
+ *   <li>"category=weather & !action=forecast"</li>
+ *   <li>"(category=weather | category=news) & !action=forecast"</li>
+ * </ul>
+ * <p>
+ * Grammar:
+ * <pre>
+ * expression  ::= orExpr
+ * orExpr      ::= andExpr ('|' andExpr)*
+ * andExpr     ::= notExpr ('&' notExpr)*
+ * notExpr     ::= '!' notExpr | primary
+ * primary     ::= '(' expression ')' | comparison
+ * comparison  ::= IDENTIFIER '=' STRING | IDENTIFIER '!=' STRING
+ * </pre>
+ */
+public class LabelExpressionParser {
```

Comment:
It would be good to have extra testing coverage on this one.

Reply:
Do you think https://github.com/wanaku-ai/wanaku/pull/632/files#diff-8fdf81d83d084e57c90f539a72e60c9782c5a55964badf8b7f4795012e12626c is not enough?